File: C++/source/Framework/Timer/TimerObjectManager.cpp
/***************************************************************************
 *   Copyright (C) 2005 Vidyasagara Guntaka                                *
 *   All rights reserved.                                                  *
 *   Author : Amr Sabaa                                                    *
 ***************************************************************************/

#include "Framework/Timer/TimerObjectManager.h"
#include "Framework/Timer/TimerSignalObjectManager.h"
#include "Framework/Utils/AssertUtils.h"

namespace WaveNs
{

TimerObjectManager::TimerObjectManager ()
    : WaveLocalObjectManager ("Framework Timer")
{
    m_pTimerWorker = new TimerWorker (this);
}

TimerObjectManager *TimerObjectManager::getInstance ()
{
    static TimerObjectManager *pTimerObjectManager = NULL;

    if (NULL == pTimerObjectManager)
    {
        pTimerObjectManager = new TimerObjectManager ();
        WaveNs::waveAssert (NULL != pTimerObjectManager, __FILE__, __LINE__);
    }

    return (pTimerObjectManager);
}

TimerObjectManager::~TimerObjectManager ()
{
    if (NULL != m_pTimerWorker)
    {
        delete m_pTimerWorker;
    }
}

WaveServiceId TimerObjectManager::getWaveServiceId ()
{
    return ((getInstance ())->getServiceId ());
}


void TimerObjectManager::disable (WaveAsynchronousContext *pWaveAsynchronousContext)
{

    trace (TRACE_LEVEL_DEVEL, "TimerObjectManager::disable : Entering ...");

    pWaveAsynchronousContext->setCompletionStatus (WAVE_MESSAGE_SUCCESS);
    pWaveAsynchronousContext->callback ();

    m_pTimerWorker->deleteAllTimers ();
}

}

File: Java/source/com/CxWave/Wave/Framework/Timer/TimerObjectManager.java
/***********************************************************************************************************
 * Copyright (C) 2015-2016 Vidyasagara Guntaka * All rights reserved. * Author : Vidyasagara Reddy Guntaka *
 ***********************************************************************************************************/

package com.CxWave.Wave.Framework.Timer;

import com.CxWave.Wave.Framework.ObjectModel.WaveLocalObjectManager;
import com.CxWave.Wave.Framework.ObjectModel.Annotations.NativeService;
import com.CxWave.Wave.Framework.ObjectModel.Annotations.ObjectManagerPriority;
import com.CxWave.Wave.Framework.ObjectModel.Annotations.PrePhaseService;
import com.CxWave.Wave.Framework.Type.WaveServiceId;
import com.CxWave.Wave.Framework.Utils.Assert.WaveAssertUtils;
import com.CxWave.Wave.Resources.ResourceEnums.WaveObjectManagerPriority;

@PrePhaseService
@NativeService
@ObjectManagerPriority (WaveObjectManagerPriority.WAVE_OBJECT_MANAGER_PRIORITY_TIMER)
public class TimerObjectManager extends WaveLocalObjectManager
{
    private static TimerObjectManager s_timerObjectManager = null;

    private TimerObjectManager ()
    {
        super (getServiceName ());
    }

    public static String getServiceName ()
    {
        return ("Framework Timer");
    }

    public static TimerObjectManager getInstance ()
    {
        if (null == s_timerObjectManager)
        {
            s_timerObjectManager = new TimerObjectManager ();
        }

        WaveAssertUtils.waveAssert (null != s_timerObjectManager);

        return (s_timerObjectManager);
    }

    public static WaveServiceId getWaveServiceId ()
    {
        return ((getInstance ()).getServiceId ());
    }
}
